client/my-sites/site-settings/jetpack-credentials/index.jsx

```diff
@@ -0,0 +1,313 @@
+			autoConfigStatus,
+			hasMainCredentials, // eslint-disable-line no-shadow
+			isPressable,
+			isRewindActive, // eslint-disable-line no-shadow
+			translate,
+		} = this.props;
+
+		const { setupStep } = this.state;
+
+		const autoConfigIdle = ( 'requesting' === autoConfigStatus || 'waiting' === autoConfigStatus );
+
+		const pressableConfigureFlow = (
+			<div className="jetpack-credentials__pressable-config-flow">
+				{ 1 === setupStep && this.renderSetupStart() }
+				{ 2 === setupStep && autoConfigIdle && this.renderSetupTos() }
+				{ 'success' === autoConfigStatus && this.renderFormFoldable() }
+				{ hasMainCredentials && this.renderFormFoldable() }
+			</div>
+		);
+
+		const selfHostedConfigureFlow = (
+			<div className="jetpack-credentials__self-hosted-config-flow">
+				{ 1 === setupStep && this.renderSetupStart() }
+				{ 2 === setupStep && this.renderSetupTos() }
+				{ 3 === setupStep && (
+					<CompactCard>
+						{ this.renderForm() }
+					</CompactCard>
+				) }
+				{ 4 === setupStep && this.renderFormFoldable() }
+			</div>
+		);
+
+		return (
+			<div className="jetpack-credentials">
+				<QueryRewindStatus siteId={ this.props.siteId } />
+				<QueryJetpackCredentials siteId={ this.props.siteId } />
+				{ isRewindActive && (
+					<CompactCard className="jetpack-credentials__header">
+						<span>{ translate( 'Backups and restores' ) }</span>
+							{ hasMainCredentials && (
+								<span className="jetpack-credentials__connected">
+									<Gridicon icon="checkmark" size={ 18 } className="jetpack-credentials__connected-checkmark" />
+									{ translate( 'Connected' ) }
+								</span>
+							) }
+					</CompactCard>
+				) }
+				{ isRewindActive && ! hasMainCredentials && isPressable && pressableConfigureFlow }
+				{ isRewindActive && ! hasMainCredentials && ! isPressable && selfHostedConfigureFlow }
+				{ isRewindActive && hasMainCredentials && this.renderFormFoldable() }
+				{ isRewindActive && 4 !== setupStep && ! hasMainCredentials && (
+					<CompactCard className="jetpack-credentials__footer">
+						<a
+							onClick={ this.togglePopover }
+							ref="popoverLink"
+							className="jetpack-credentials__footer-popover-link"
+						>
+							<Gridicon icon="help" size={ 18 } className="jetpack-credentials__footer-popover-icon" />
+							{
+								3 === setupStep
+									? translate( 'Need help finding your site\'s server credentials?' )
+									: translate( 'Why do I need this?' )
+							}
+						</a>
+						<Popover
+							context={ this.refs && this.refs.popoverLink }
+							isVisible={ get( this.state, 'showPopover', false ) }
+							onClose={ this.togglePopover }
+							className="jetpack-credentials__footer-popover"
+							position="top"
+						>
+							{ translate(
+								'These credentials are used to perform automatic actions ' +
+								'on your server including backups and restores.'
+							) }
+						</Popover>
+					</CompactCard>
+				) }
+			</div>
+		);
+	}
+}
```

Comment:
notice how in this function we have several different places where we are checking for the state of the _machine_.

is it possible that we've discovered a _page_ or _section_ "wrapper" component which decodes the machine and display the appropriate _sub-page_ for each state?

it would surprise me if things didn't become simpler if we extracted these pages into separate pages and were left with one big dispatch…

```js
render() {
	const { setupStep } = this.props;

	switch ( setupStep ) {
		case 1:
			return <RewindStarter beginProcess={ this.nextStep } />;

		case 2:
			return <TOSConfirmation confirmTOS={ this.nextStep } />;

		case 3:
			return <CredentialsForm credentials={ … } submitCredentials={ this.submitCredentials } />;

		case 4:
			return <CredentialsList />;
	}

	return <UnknownStep />;
}
```

If we can continue to extract these things then we'll be left with much less to consider when we're working on any given piece of code, plus we can remove transition code we thought we needed. For example, we can handle all of the coordination in one function and spit out the Rewind state needed for this component - if we update the API, we can remove that translator.

```js
const getRewindState = ( state, siteId ) => {
	const rewindStatus = getRewindStatus( state, siteId );

	if ( rewindStatus === null ) {
		return { state: 'rewindDeactivated' };
	}

	const credentials = getCredentials( state, siteId );

	if ( ! credentials.length ) {
		return { state: 'waitingOnCredentials' };
	}

	return { state: 'rewindActive' }
}
```

^^^ that's not correct, but should convey the idea of containing and simplifying state as we get closer to the customer

Reply:
Well said Dennis. This section could be clarified a bit Rich.

For example, what gets rendered here if rewind is not active?

Both `renderConfiguredView` and `jetpack-credentials__header` hinge on the same logic. Could they be combined into a separate component, rather than an internal render method?

Both `CredentialsSetupFlow` and the popover bit hinge on the same logic, so they could be combined within the same component?